lib/cmsis_rtos_v1/cmsis_semaphore.c

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_structs.h>
+#include <cmsis_os.h>
+
+K_MEM_SLAB_DEFINE(cmsis_semaphore_slab, sizeof(struct k_sem),
+		CONFIG_CMSIS_SEMAPHORE_MAX_COUNT, 4);
+
+/**
+ * @brief Create and Initialize a semaphore object.
+ */
+osSemaphoreId osSemaphoreCreate(const osSemaphoreDef_t *semaphore_def,
+				int32_t count)
+{
+	struct k_sem *semaphore;
+
+	if (semaphore_def == NULL) {
+		return NULL;
+	}
+
+	if (_is_in_isr()) {
+		return NULL;
+	}
+
+	if (k_mem_slab_alloc(&cmsis_semaphore_slab,
+				(void **)&semaphore, 100) == 0) {
+		memset(semaphore, 0, sizeof(struct k_sem));
+	} else {
+		return NULL;
+	}
+
+	k_sem_init(semaphore, count, count);
+
+	return (osSemaphoreId)semaphore;
+}
+
+/**
+ * @brief Wait until a semaphore becomes available.
+ */
+int32_t osSemaphoreWait(osSemaphoreId semaphore_id, uint32_t timeout)
+{
+	struct k_sem *semaphore = (struct k_sem *) semaphore_id;
+	int status;
+
+	if (semaphore_id == NULL) {
+		return -1;
+	}
+
+	if (_is_in_isr()) {
+		return -1;
+	}
+
+	if (timeout == osWaitForever) {
+		status = k_sem_take(semaphore, K_FOREVER);
+	} else if (timeout == 0) {
+		status = k_sem_take(semaphore, K_NO_WAIT);
+	} else {
+		status = k_sem_take(semaphore, timeout);
+	}
+
+	/* If k_sem_take is successful, return the number of available
+	 * tokens.
+	 */
+	if (status == 0) {
+		return k_sem_count_get(semaphore);
+	}
+
+	return -1;
```

Comment:
I'm not sure whether returning -1 here is right - could you check? In CMSIS doc, it says to return

> -1 in case of incorrect parameters

which is okay for the first two checks at the start; but in the case that the semaphore waiting period timed out, `k_sem_take` would return non-zero (EAGAIN) to `status`, indicating unavailability, and that should mean returning 0 rather -1 from `osSemaphoreWait`, since acc to doc again,

> The return value indicates the number of available tokens (the semaphore count value). If 0 is returned, then no semaphore was available.

Reply:
That's correct. status will be 0 only if the semaphore was taken. So, in case of timeout, i should be returning 0 (indicating unavailability) instead of -1 (which is an error). Thanks for the catch.